### ISSUE TITLE
System internal readiness probe overrides my custom probe and blocks incoming traffic when it triggers

### ISSUE THREAD

**Issue by vegarnilsen** [microsoft/azure-container-apps#687]
> Please provide us with the following information:
> ---------------------------------------------------------------

### This issue is a: (mark with an x)
- [x] bug report -> please search issues before submitting
- [ ] documentation issue or request
- [ ] regression (a behavior that used to work and stopped in a new release)

### Issue description
I'm running a wordpress site as a set of individual container apps, where the relevant apps for this issue is a Varnish container that receives incoming traffic from Application Gateway, and then a container running Apache with mod-php for Wordpress itself. On several occasions the site has been down, because a readiness probe has marked the container app as PodUnhealthy, and then no traffic reaches the containers until I manually restart the revision. (See log of this behaviour below.)

I've configured custom probes like this (via Bicep):

```bicep
          probes: [
            {
              type: 'Liveness'
              failureThreshold: 3
              initialDelaySeconds: 10
              periodSeconds: 10
              successThreshold: 1
              httpGet: {
                port: 80
                httpHeaders: [
                  {
                    name: 'Host'
                    value: 'wordpress.internal.${cae.properties.defaultDomain}'
                  }
                ]
                path: '/readme.html'
                scheme: 'HTTP'
              }
              timeoutSeconds: 1
            }
            {
              type: 'Readiness'
              failureThreshold: 3
              initialDelaySeconds: 10
              periodSeconds: 10
              successThreshold: 1
              httpGet: {
                port: 80
                httpHeaders: [
                  {
                    name: 'Host'
                    value: 'wordpress.internal.${cae.properties.defaultDomain}'
                  }
                ]
                path: '/readme.html'
                scheme: 'HTTP'
              }
              timeoutSeconds: 1
            }
          ]
```

Here I'm asking for `/readme.html` because I just want to know that the container app is up and Apache is responding to incoming requests, without getting some problem due to slow response of uncached content.

I've also set up similar probes for the Varnish container, and configured Varnish to respond directly on a specific path so that I monitor the health of Varnish itself, not the Wordpress site behind Varnish.

### Steps to reproduce
I'm not entirely sure how to reproduce this problem since it seems to happen somewhat randomly, but it usually happens at night (local time) or during weekends.

Once this problem happened I solved it by restarting the revision like this:

```powershell
PS > $wordpress_rev = Get-AzContainerAppRevision -ResourceGroupName $rgName -ContainerAppName wordpress | where Active
PS > Restart-AzContainerAppRevision -ResourceGroupName $rgName -ContainerAppName wordpress -RevisionName $wordpress_rev.Name
```

**Expected behavior** 

I would expect that the system uses my health probes to determine if the container is responding or not, and if I use the same check type for liveness and readiness I would expect the liveness check to be prioritised in case of failures so the container is automatically restarted on repeated failures.

**Actual behavior** 

A readiness probe that I haven't configured is triggering, putting the container into status PodUnhealthy forever (I've seen the count reach 25000 failed checks during a weekend failure) and nothing else happening to that container.

**Screenshots**         

Entire log for a replica that failed as described above

```csv
"TimeGenerated [UTC]","Reason_s","Count_d","Log_s"
"3/22/2023, 6:58:36.239 AM",PodUnhealthy,867,"Readiness probe failed: stat: cannot statx '/ready': No such file or directory"
"3/22/2023, 6:58:36.239 AM",PodUnhealthy,867,"Readiness probe failed: stat: cannot statx '/ready': No such file or directory"
"3/22/2023, 6:53:35.786 AM",PodUnhealthy,833,"Readiness probe failed: stat: cannot statx '/ready': No such file or directory"
"3/22/2023, 6:53:35.786 AM",PodUnhealthy,833,"Readiness probe failed: stat: cannot statx '/ready': No such file or directory"
"3/22/2023, 6:48:35.746 AM",PodUnhealthy,799,"Readiness probe failed: stat: cannot statx '/ready': No such file or directory"
"3/22/2023, 6:48:35.746 AM",PodUnhealthy,799,"Readiness probe failed: stat: cannot statx '/ready': No such file or directory"
"3/22/2023, 6:43:36.093 AM",PodUnhealthy,765,"Readiness probe failed: stat: cannot statx '/ready': No such file or directory"
"3/22/2023, 6:43:36.093 AM",PodUnhealthy,765,"Readiness probe failed: stat: cannot statx '/ready': No such file or directory"
"3/22/2023, 6:38:35.828 AM",PodUnhealthy,731,"Readiness probe failed: stat: cannot statx '/ready': No such file or directory"
"3/22/2023, 6:38:35.828 AM",PodUnhealthy,731,"Readiness probe failed: stat: cannot statx '/ready': No such file or directory"
"3/22/2023, 6:33:35.883 AM",PodUnhealthy,696,"Readiness probe failed: stat: cannot statx '/ready': No such file or directory"
"3/22/2023, 6:33:35.883 AM",PodUnhealthy,696,"Readiness probe failed: stat: cannot statx '/ready': No such file or directory"
"3/22/2023, 6:28:35.739 AM",PodUnhealthy,662,"Readiness probe failed: stat: cannot statx '/ready': No such file or directory"
"3/22/2023, 6:28:35.739 AM",PodUnhealthy,662,"Readiness probe failed: stat: cannot statx '/ready': No such file or directory"
"3/22/2023, 6:23:35.916 AM",PodUnhealthy,629,"Readiness probe failed: stat: cannot statx '/ready': No such file or directory"
"3/22/2023, 6:23:35.916 AM",PodUnhealthy,629,"Readiness probe failed: stat: cannot statx '/ready': No such file or directory"
"3/22/2023, 6:18:35.926 AM",PodUnhealthy,595,"Readiness probe failed: stat: cannot statx '/ready': No such file or directory"
"3/22/2023, 6:18:35.926 AM",PodUnhealthy,595,"Readiness probe failed: stat: cannot statx '/ready': No such file or directory"
"3/22/2023, 6:13:35.839 AM",PodUnhealthy,561,"Readiness probe failed: stat: cannot statx '/ready': No such file or directory"
"3/22/2023, 6:13:35.839 AM",PodUnhealthy,561,"Readiness probe failed: stat: cannot statx '/ready': No such file or directory"
"3/22/2023, 6:08:35.780 AM",PodUnhealthy,527,"Readiness probe failed: stat: cannot statx '/ready': No such file or directory"
"3/22/2023, 6:08:35.780 AM",PodUnhealthy,527,"Readiness probe failed: stat: cannot statx '/ready': No such file or directory"
"3/22/2023, 6:03:35.739 AM",PodUnhealthy,493,"Readiness probe failed: stat: cannot statx '/ready': No such file or directory"
"3/22/2023, 6:03:35.739 AM",PodUnhealthy,493,"Readiness probe failed: stat: cannot statx '/ready': No such file or directory"
"3/22/2023, 5:58:35.842 AM",PodUnhealthy,459,"Readiness probe failed: stat: cannot statx '/ready': No such file or directory"
"3/22/2023, 5:58:35.842 AM",PodUnhealthy,459,"Readiness probe failed: stat: cannot statx '/ready': No such file or directory"
"3/22/2023, 5:53:35.876 AM",PodUnhealthy,425,"Readiness probe failed: stat: cannot statx '/ready': No such file or directory"
"3/22/2023, 5:53:35.876 AM",PodUnhealthy,425,"Readiness probe failed: stat: cannot statx '/ready': No such file or directory"
"3/22/2023, 5:48:35.666 AM",PodUnhealthy,391,"Readiness probe failed: stat: cannot statx '/ready': No such file or directory"
"3/22/2023, 5:48:35.666 AM",PodUnhealthy,391,"Readiness probe failed: stat: cannot statx '/ready': No such file or directory"
"3/22/2023, 5:43:36.125 AM",PodUnhealthy,357,"Readiness probe failed: stat: cannot statx '/ready': No such file or directory"
"3/22/2023, 5:43:36.125 AM",PodUnhealthy,357,"Readiness probe failed: stat: cannot statx '/ready': No such file or directory"
"3/22/2023, 5:38:35.816 AM",PodUnhealthy,323,"Readiness probe failed: stat: cannot statx '/ready': No such file or directory"
"3/22/2023, 5:38:35.816 AM",PodUnhealthy,323,"Readiness probe failed: stat: cannot statx '/ready': No such file or directory"
"3/22/2023, 5:33:35.912 AM",PodUnhealthy,289,"Readiness probe failed: stat: cannot statx '/ready': No such file or directory"
"3/22/2023, 5:33:35.912 AM",PodUnhealthy,289,"Readiness probe failed: stat: cannot statx '/ready': No such file or directory"
"3/22/2023, 5:28:35.630 AM",PodUnhealthy,255,"Readiness probe failed: stat: cannot statx '/ready': No such file or directory"
"3/22/2023, 5:28:35.630 AM",PodUnhealthy,255,"Readiness probe failed: stat: cannot statx '/ready': No such file or directory"
"3/22/2023, 5:23:35.749 AM",PodUnhealthy,221,"Readiness probe failed: stat: cannot statx '/ready': No such file or directory"
"3/22/2023, 5:23:35.749 AM",PodUnhealthy,221,"Readiness probe failed: stat: cannot statx '/ready': No such file or directory"
"3/22/2023, 5:18:36.348 AM",PodUnhealthy,187,"Readiness probe failed: stat: cannot statx '/ready': No such file or directory"
"3/22/2023, 5:18:36.348 AM",PodUnhealthy,187,"Readiness probe failed: stat: cannot statx '/ready': No such file or directory"
"3/22/2023, 5:13:36.341 AM",PodUnhealthy,153,"Readiness probe failed: stat: cannot statx '/ready': No such file or directory"
"3/22/2023, 5:13:36.341 AM",PodUnhealthy,153,"Readiness probe failed: stat: cannot statx '/ready': No such file or directory"
"3/22/2023, 5:08:35.754 AM",PodUnhealthy,119,"Readiness probe failed: stat: cannot statx '/ready': No such file or directory"
"3/22/2023, 5:08:35.754 AM",PodUnhealthy,119,"Readiness probe failed: stat: cannot statx '/ready': No such file or directory"
"3/22/2023, 5:03:35.623 AM",PodUnhealthy,85,"Readiness probe failed: stat: cannot statx '/ready': No such file or directory"
"3/22/2023, 5:03:35.623 AM",PodUnhealthy,85,"Readiness probe failed: stat: cannot statx '/ready': No such file or directory"
"3/22/2023, 4:58:35.852 AM",PodUnhealthy,51,"Readiness probe failed: stat: cannot statx '/ready': No such file or directory"
"3/22/2023, 4:58:35.852 AM",PodUnhealthy,51,"Readiness probe failed: stat: cannot statx '/ready': No such file or directory"
"3/22/2023, 4:53:36.159 AM",PodUnhealthy,17,"Readiness probe failed: stat: cannot statx '/ready': No such file or directory"
"3/22/2023, 4:53:36.159 AM",PodUnhealthy,17,"Readiness probe failed: stat: cannot statx '/ready': No such file or directory"
"3/22/2023, 4:52:05.771 AM",PodUnhealthy,7,"Readiness probe failed: stat: cannot statx '/ready': No such file or directory"
"3/22/2023, 4:52:05.771 AM",PodUnhealthy,7,"Readiness probe failed: stat: cannot statx '/ready': No such file or directory"
"3/22/2023, 4:51:57.552 AM",PodUnhealthy,6,"Readiness probe failed: stat: cannot statx '/ready': No such file or directory"
"3/22/2023, 4:50:00.794 AM",ContainerStarted,1,"Started container wordpress"
"3/22/2023, 4:50:00.794 AM",ContainerStarted,1,"Started container wordpress"
"3/22/2023, 4:50:00.794 AM",ContainerCreated,1,"Created container wordpress"
"3/22/2023, 4:50:00.794 AM",ImagePulled,1,"Successfully pulled image ""****.azurecr.io/****/dev_image:sha-954e5d2"" in 1m21.250379281s"
"3/22/2023, 4:50:00.794 AM",ImagePulled,1,"Successfully pulled image ""****.azurecr.io/****/dev_image:sha-954e5d2"" in 1m21.250379281s"
"3/22/2023, 4:48:44.449 AM",PullingImage,1,"Pulling image ""****.azurecr.io/****/dev_image:sha-954e5d2"""
"3/22/2023, 4:48:44.449 AM",PullingImage,1,"Pulling image ""****.azurecr.io/****/dev_image:sha-954e5d2"""
"3/22/2023, 4:48:43.530 AM",ContainerStarted,1,"Started container k8se-localproxy"
"3/22/2023, 4:48:43.530 AM",ContainerStarted,1,"Started container k8se-localproxy"
"3/22/2023, 4:48:42.402 AM",ContainerCreated,1,"Created container k8se-localproxy"
"3/22/2023, 4:48:42.402 AM",ContainerCreated,1,"Created container k8se-localproxy"
"3/22/2023, 4:48:42.402 AM",ImagePulled,1,"Container image ""mcr.microsoft.com/k8se/proxy:main-426eed79"" already present on machine"
"3/22/2023, 4:48:42.402 AM",ImagePulled,1,"Container image ""mcr.microsoft.com/k8se/proxy:main-426eed79"" already present on machine"
"3/22/2023, 4:48:40.535 AM",ContainerStarted,1,"Started container k8se-envoysc"
"3/22/2023, 4:48:40.535 AM",ContainerStarted,1,"Started container k8se-envoysc"
"3/22/2023, 4:48:39.477 AM",ContainerCreated,1,"Created container k8se-envoysc"
"3/22/2023, 4:48:39.477 AM",ContainerCreated,1,"Created container k8se-envoysc"
"3/22/2023, 4:48:39.477 AM",ImagePulled,1,"Container image ""mcr.microsoft.com/k8se/envoy:main-426eed79"" already present on machine"
"3/22/2023, 4:48:39.477 AM",ImagePulled,1,"Container image ""mcr.microsoft.com/k8se/envoy:main-426eed79"" already present on machine"
"3/22/2023, 4:48:36.401 AM",FailedMount,2,"MountVolume.SetUp failed for volume ""identity-config"" : failed to sync configmap cache: timed out waiting for the condition"
"3/22/2023, 4:48:36.401 AM",FailedMount,2,"MountVolume.SetUp failed for volume ""identity-config"" : failed to sync configmap cache: timed out waiting for the condition"
"3/22/2023, 4:48:36.401 AM",FailedMount,2,"MountVolume.SetUp failed for volume ""identity-secret"" : failed to sync secret cache: timed out waiting for the condition"
"3/22/2023, 4:48:36.401 AM",FailedMount,2,"MountVolume.SetUp failed for volume ""identity-secret"" : failed to sync secret cache: timed out waiting for the condition"
"3/22/2023, 4:48:32.354 AM",FailedMount,1,"MountVolume.SetUp failed for volume ""identity-config"" : failed to sync configmap cache: timed out waiting for the condition"
"3/22/2023, 4:48:32.354 AM",FailedMount,1,"MountVolume.SetUp failed for volume ""identity-secret"" : failed to sync secret cache: timed out waiting for the condition"
"3/22/2023, 4:48:31.659 AM",AssigningPod,0,"Successfully assigned k8se-apps/wordpress--5smfdmb-7d48968b44-ckqtc to aks-agentpool-41787472-vmss000001"
```

### Additional context
This happens in both our development and production environments, and this web site gets enough traffic (in production) that it seems unlikely to be because of low activity. Regardless of activity we've also specified that we want a minimum of one replica running at all times for these containers.

**Comment by torosent** [comment]
Hi All, 
Hi, can you send your sub id, env name and apps names to `acasupport at microsoft dot com`?

**Comment by torosent** [comment]
Also, we have a known issue that when you are setting the readiness and liveness probes and not the startup probe, we are not removing the startup probe. That might be the issue here. We have a bug fix that will rollout soon but in the meantime, try to setup the startup probe.

**Comment by vegarnilsen** [comment]
> Also, we have a known issue that when you are setting the readiness and liveness probes and not the startup probe, we are not removing the startup probe. That might be the issue here. We have a bug fix that will rollout soon but in the meantime, try to setup the startup probe.

The probe setup I posted here was for the wordpress container, for the varnish container I've configured the startup probe in addition to liveness and readiness, and see the same behaviour there as well.

In both cases the containers work for a while, several days in some instances, before the readiness probe triggers for some reason and stops traffic reaching the container.

I will add the startup probe to the wordpress container just to avoid the issue you mentioned.

**Comment by vegarnilsen** [comment]
> Hi All, Hi, can you send your sub id, env name and apps names to `acasupport at microsoft dot com`?

I've sent this now.

**Comment by passarela** [comment]
The behavior in the environment is similar. I have readiness probe configured:
![chrome_5yATX87xWq](https://user-images.githubusercontent.com/72925771/227193869-e14bc568-e070-4850-a4dc-ffcaae49efad.png)

However the /ready error logs continue to appear. I could not notice a default behavior happens several times an hour.
![chrome_JZqNMtqeHr](https://user-images.githubusercontent.com/72925771/227193937-98e8a615-8a8d-4c71-a79b-c0579efe25c4.png)
![ShareX_7A8qVwxd7b](https://user-images.githubusercontent.com/72925771/227193964-1cd0daa3-405d-4bc6-a6f7-cced3128315f.png)


I emphasize that the configured probe must not check the /ready directory because it does not exist in my application.

It should respect my configured probe /api/health which is accessible and returns http 200

**Comment by vegarnilsen** [comment]
In case this problem was caused by some change to a lower level in the stack, we've seen this same type of crash since March 8th. (At the time we didn't have any custom health probes defined, and the log was the same "Readiness probe failed: stat: cannot statx '/ready'…" message.)

**Comment by sanchitmehta** [comment]
The System logs that you see with probes failing for `http://<ip>:23045` and the message `prob failed stat cannot statx ‘/ready’: No such file or directory` are for Container Apps Platform containers.

These containers are injected as sidecars to your app to set up routing. These logs can be safely ignored.

There is a bug on our side which exposes these logs to your Log Analytics/Azure Monitor workspace. We are working on fixing this and should have a patch out soon.

**Comment by vegarnilsen** [comment]
> The System logs that you see with probes failing for `http://<ip>:23045` and the message `prob failed stat cannot statx ‘/ready’: No such file or directory` are for Container Apps Platform containers.
> 
> These containers are injected as sidecars to your app to set up routing. These logs can be safely ignored.
> 
> There is a bug on our side which exposes these logs to your Log Analytics/Azure Monitor workspace. We are working on fixing this and should have a patch out soon.

So, if I can safely ignore these logs, why is it entirely consistent that my site is down in the specific case where that message has been repeated a number of times? This has been the case every time we've had downtime since March 8th, so it seems like there's more than just a cosmetic issue here.

Since these sidecar containers are tasked with routing, could problems there lead to downtime between Application Gateway and my containers?

**Comment by vegarnilsen** [comment]
It looks like there were two problems here, one was the sidecar readiness probe, and then the other was an update to the underlying Kubernetes infrastructure conflicting with some changes I'd done to `/etc/resolv.conf` to let the container resolve external DNS names. (`options ndots:1`)

After I rebuilt our containers without that modification I haven't seen any more such runaway readiness probes, nor any downtime. Any potential container problems have been caught by the health probes I configured for our containers, and they've been restarted as expected.
